app/adapters/signal_adapter/inbound.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+module SignalAdapter
+  UNKNOWN_CONTRIBUTOR = :unknown_contributor
+  UNKNOWN_CONTENT = :unknown_content
+  CONNECT = :connect
+
+  class Inbound
+    UNKNOWN_CONTENT_KEYS = %w[mentions contacts reaction sticker].freeze
+    SUPPORTED_ATTACHMENT_TYPES = %w[image/jpg image/jpeg image/png image/gif audio/oog audio/aac audio/mp4].freeze
+
+    attr_reader :sender, :text, :message
+
+    def initialize
+      @callbacks = {}
+    end
+
+    def on(callback, &block)
+      @callbacks[callback] = block
+    end
+
+    def consume(signal_message)
+      signal_message = signal_message.with_indifferent_access
+
+      @sender = initialize_sender(signal_message)
+      return unless @sender
+
+      @message = initialize_message(signal_message)
+      return unless @message
+
+      files = initialize_files(signal_message)
+      @message.files = files
+
+      has_content = @message.text || @message.files.any? || @message.unknown_content
+      return unless has_content
+
+      yield(@message) if block_given?
```

Comment:
I know I was the first to introduce this "style" of coding the adapters, however, I believe it's quite complex and hard to read. I'm a little sorry to not have it refactored earlier and see this code being repeated for other adapters :see_no_evil: :sweat_smile:

Reply:
Are you talking about the general structure of adapters or about the yield/block to handle incoming message? If the latter is the case, maybe we could handle messages the same way we handle any other event? That might be less confusing…

```ruby
adapter.on(SignalAdapter::MESSAGE) do |message|
  ...
end
```

Reply:
I think the yield/block has quite a nice API, I think that all these private methods have quite some dependencies on each other, ie. they have side effects and must run in a specific order. This is quite bad I'd say.